core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationsToSubqueries.java

```diff
@@ -0,0 +1,235 @@
+            Map<Symbol, Aggregation> aggregations = aggregationsByArgumentsList.get(i);
+            AggregationNode subAggregationNode = buildSubAggregation(aggregationNode, aggregations, assignments, context);
+
+            if (right == null) {
+                right = subAggregationNode;
+                rightJoinSymbols = subAggregationNode.getGroupingKeys();
+            }
+            else {
+                right = buildJoin(subAggregationNode, subAggregationNode.getGroupingKeys(), right, rightJoinSymbols, context);
+            }
+        }
+
+        // the first aggregation is the left side of the top join
+        AggregationNode left = buildSubAggregation(aggregationNode, aggregationsByArgumentsList.get(0), assignments, context);
+
+        for (int i = 0; i < left.getGroupingKeys().size(); i++) {
+            assignments.put(aggregationNode.getGroupingKeys().get(i), left.getGroupingKeys().get(i).toSymbolReference());
+        }
+        JoinNode topJoin = buildJoin(left, left.getGroupingKeys(), right, rightJoinSymbols, context);
+        ProjectNode result = new ProjectNode(aggregationNode.getId(), topJoin, assignments.build());
+        return Result.ofPlanNode(result);
+    }
+
+    private AggregationNode buildSubAggregation(AggregationNode aggregationNode, Map<Symbol, Aggregation> aggregations, Assignments.Builder assignments, Context context)
+    {
+        List<Symbol> originalAggregationOutputSymbols = ImmutableList.copyOf(aggregations.keySet());
+        // copy the plan so that both plan node ids and symbols are not duplicated between sub aggregations
+        NodeAndMappings copied = PlanCopier.copyPlan(
+                AggregationNode.builderFrom(aggregationNode).setAggregations(aggregations).build(),
+                originalAggregationOutputSymbols,
+                context.getSymbolAllocator(),
+                context.getIdAllocator(),
+                context.getLookup());
+        AggregationNode subAggregationNode = (AggregationNode) copied.getNode();
+        // add the mapping from the new output symbols to original ones
+        for (int i = 0; i < originalAggregationOutputSymbols.size(); i++) {
+            assignments.put(originalAggregationOutputSymbols.get(i), copied.getFields().get(i).toSymbolReference());
+        }
+        return subAggregationNode;
+    }
+
+    private JoinNode buildJoin(PlanNode left, List<Symbol> leftJoinSymbols, PlanNode right, List<Symbol> rightJoinSymbols, Context context)
+    {
+        checkArgument(leftJoinSymbols.size() == rightJoinSymbols.size());
+        List<EquiJoinClause> criteria = IntStream.range(0, leftJoinSymbols.size())
+                .mapToObj(i -> new EquiJoinClause(leftJoinSymbols.get(i), rightJoinSymbols.get(i)))
+                .collect(toImmutableList());
+
+        // TODO: we dont need dynamic filters for this join at all. We could add skipDf field to the JoinNode and make use of it in PredicatePushDown
```

Comment:
I didn't get what this TODO means, why would we "add skipDf field to the JoinNode and make use of it in PredicatePushDown" ? If there is no DF to be pushed down, then PredicatePushDown will do nothing.

Reply:
The join created here joins multiple sources that are guaranteed to have the same values in the columns that are part of join condition. This makes DF not beneficial at all so it just adds overhead. We know this here but this knowledge is lost in the `PredicatePushDown` where the DF is created

Reply:
okay, i think see the point now